### PR TITLE
Feature/env file

### PR DESCRIPTION
**Pull request by cw-andrews**
Implementation of .env files, accompanying tests, and appropriate changes to the docs.

### CODE REVIEW

File: docs/guide/environment.rst [hunk 1]
```diff
@@ -23,25 +23,76 @@ simply to use the optional ``--env`` argument when running bonobo from the shell
...
+
+Examples
+::::::::
 
 The Examples below demonstrate setting one or multiple variables using both of these methods:
 
 .. code-block:: bash
 
-    # Using one environment variable via --env flag:
+    # Using one environment variable via --env and --defualt-env flags:
```
Comment (hartym) [inline docs/guide/environment.rst]: Typo here

File: docs/guide/environment.rst [hunk 2]
```diff
@@ -23,25 +23,76 @@ simply to use the optional ``--env`` argument when running bonobo from the shell
...
     bonobo run csvsanitizer --env SECRET_TOKEN=secret123
+    bonobo run csvsanitizer --defaul-env SECRET_TOKEN=secret123
```
Comment (hartym) [inline docs/guide/environment.rst]: Typo here too.

File: docs/guide/environment.rst [hunk 3]
```diff
@@ -23,25 +23,76 @@ simply to use the optional ``--env`` argument when running bonobo from the shell
...
     SECRET_TOKEN=secret123 bonobo run csvsanitizer
 
-    # Using multiple environment variables inline (bash only):
+    # Using multiple environment variables inline (bash-like shells only):
     SRC_FILE=inventory.txt DST_FILE=inventory_processed.csv bonobo run csvsanitizer
-    
-*Though not-yet implemented, the bonobo roadmap includes implementing environment / .env files as well.*
+
+    # Using an env file for default env values:
+    bonobo run csvsanitizer --default-env-file .default_env
```
Comment (hartym) [inline docs/guide/environment.rst]: I'd prefer to use ".env" here, as it's the usual behaviour of .env with honcho, foreman, etc.

File: docs/guide/environment.rst [hunk 4]
```diff
@@ -23,25 +23,76 @@ simply to use the optional ``--env`` argument when running bonobo from the shell
...
+
+    # Using an env file for env values:
+    bonobo run csvsanitizer --env-file '.env'
```
Comment (hartym) [inline docs/guide/environment.rst]: Let's use .env.private (or .env.local) here

File: requirements.txt [hunk 5]
```diff
@@ -2,13 +2,15 @@
...
 chardet==3.0.4
+click==6.7
 colorama==0.3.9
 fs==2.0.11
 idna==2.6
 packaging==16.8
 pbr==3.1.1
 psutil==5.3.1
 pyparsing==2.2.0
+python-dotenv==0.7.1
```
Comment (hartym) [inline requirements.txt]: The requirements are managed by Projectfile / edgy.project, I'll handle this if you prefer (requirements*.txt are rewritten automatically on a regular basis).

**Comment by hartym** [comment]
Also, it looks that your patch borke the windows CI ;)

File: requirements.txt (hunk 5 above)
Comment (cw-andrews) [reply requirements.txt]: If I am right in thinking that I just need to add the requirement as another element in this list then I think I can swing it :). If not then maybe I will take you up on that.

screenshot from 2017-10-15 17-05-51

**Comment by cw-andrews** [comment]
@hartym Ya, I looked at the Windows CI console output and it looked like it just wasn't installing python-dotenv - probably because of how I messed-up the requirements part of adding a dependency.

File: requirements.txt (hunk 5 above)
Comment (cw-andrews) [reply requirements.txt]: @hartym Am I right about what I have to change? If so I will make the change now.

Comment (cw-andrews) [reply requirements.txt]: @hartym Actually, I now think as it is a base requirement I would have to add it right after stevedore.

Anyway let me know!

Comment (hartym) [reply requirements.txt]: Yes, you just have to add it in the "base" requirements, no need to add it also in extras. Usually, I keep the requirements sorted alphabetically, for no good reason other than being a bit maniac. Once it is done, you need either to `pip install edgy.project; edgy-project update` at the root of the project, or I can do it for you if any trouble.

**Comment by hartym** [comment]
About Windows CI, that's probably true. `edgy-project update` will generate requirements files but also and most importantly the `setup.py`.

**Comment by cw-andrews** [comment]
@hartym BTW, do you sleep? :)

File: requirements.txt (hunk 5 above)
Comment (cw-andrews) [reply requirements.txt]: It isn't, just did that. We will see soon whether everything is good with the Windows CI now. Also, are the doc changes good?

**Comment by cw-andrews** [comment]
Cool, I am sure you will let me know if anything else needs changed but it looks like everything is passing now.

**Comment by hartym** [comment]
Sorry for not getting back to you sooner (work...). I'll merge it for now in develop, so we can test further in different contexts and see the documentation updates. I'll let you know if I see strange things around. Thanks!

**Pull request merged**